### DISCUSSION TITLE
How to check if MIDI is (still) connected?

### DISCUSSION THREAD

**Discussion by FDelporte** [atsushieno/ktmidi#64] [General]
As I found out, once the connection state is OPEN, it stays like that, even when you disconnect the cable.
What would be a good way to check if the device is still connected?

```
interface MidiPort {
    val details: MidiPortDetails
    val connectionState: MidiPortConnectionState
    fun close()
}
```

**Comment by atsushieno** [comment]
You're hitting an issue that I once faced (and haven't sorted out what to do) - https://github.com/atsushieno/ktmidi/issues/43

In other words, it is an unresolved problem within the current API. It should be doable on most, if not all, target platforms though.

**Reply by FDelporte** [reply]
Maybe onEventReceived can be used to store last timestamp and if nothing is received for X seconds, it's considered to be disconnected? I'll try such an approach in my app... But that doesn't work for a MidiOutput I guess...

**Reply by atsushieno** [reply]
Thanks for the ideas, but I guess that's not going to be good enough - we connect an input device and leave it unused for a while. Hardware MIDI devices would send active sensing. Software (virtual) devices would not. We also need device change detection for output devices too.

**Comment by atsushieno** [comment]
#43 is (finally) handled (with breaking API change, though I'm sure no one has used it as none of our `MidiAccess` implementation supported it).

While exploring other x-plat MIDI access APIs I noticed that most of them do not detect device changes, especially on Windows (winmm does not support it while WinRT does). RtMidi does not either so it would not work with most of the platforms as long as we use `RtMidiAccess`.

Chromium does handle whatever Web MIDI API requires though. Not sure if it supports Windows though (I haven't tested).
